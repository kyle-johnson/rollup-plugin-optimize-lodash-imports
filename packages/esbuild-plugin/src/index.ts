import fs from "fs";

import * as acorn from "acorn";
import { Plugin } from "esbuild";
import {
  ParseFunction,
  transform,
  UNCHANGED,
} from "@optimize-lodash/transform";

const wrappedParse: ParseFunction = (code) =>
  acorn.parse(code, { ecmaVersion: "latest", sourceType: "module" });

export type PluginOptions = {
  useLodashEs?: true;
  appendDotJs?: boolean;
};

// TODO: filter https://golang.org/pkg/regexp/
export function lodashOptimizeImports({
  useLodashEs,
  appendDotJs = true,
}: PluginOptions = {}): Plugin {
  const cache = new Map<
    string,
    { input: string; output: string | UNCHANGED }
  >();

  return {
    name: "lodash-optimize-imports",
    setup(build) {
      build.onLoad({ filter: /.(js|ts|jsx|tsx)$/ }, async ({ path }) => {
        const input = await fs.promises.readFile(path, "utf8");
        const cached = cache.get(path); // TODO: unit test the cache

        if (cached && input === cached.input && cached.output === UNCHANGED) {
          return;
        }

        const result = transform({
          code: input,
          id: path,
          parse: wrappedParse,
          useLodashEs,
          appendDotJs,
        });
        if (result === UNCHANGED) {
          cache.set(path, { input, output: UNCHANGED });
          return;
        }

        const output = result.code;
        cache.set(path, { input, output });
        return { contents: output };
      });
    },
  };
}
