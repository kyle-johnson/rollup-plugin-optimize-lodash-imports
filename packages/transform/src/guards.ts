import type {
  BaseNode,
  ImportDeclaration,
  ImportSpecifier,
  Program,
} from "estree";

export function isImportDeclaration(node: BaseNode): node is ImportDeclaration {
  return node.type === "ImportDeclaration";
}

export function isProgram(node: BaseNode): node is Program {
  return node.type === "Program";
}

export function isImportSpecifierArray(
  items: ImportDeclaration["specifiers"],
): items is Array<ImportSpecifier> {
  return items.every((item) => item.type === "ImportSpecifier");
}
