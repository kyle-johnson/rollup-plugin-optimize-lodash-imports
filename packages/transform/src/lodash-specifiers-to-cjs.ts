import type { ImportSpecifier } from "estree";

/**
 * Turns a generic lodash import into a specific import using the CommonJS
 * lodash package.
 *
 * @param base "lodash" or "lodash/fp"
 * @param specifiers from an AST; assumes they are all ImportSpecifiers
 * @param appendDotJs optional, default is true; adds '.js' to the end of imports
 */
export function lodashSpecifiersToCjs(
  base: string,
  specifiers: Array<ImportSpecifier>,
  appendDotJs = true
): Array<string> {
  return specifiers.map(
    ({ imported, local }) =>
      `import ${
        imported.name !== local.name ? local.name : imported.name
      } from "${base}/${imported.name}${appendDotJs ? ".js" : ""}";`
  );
}
