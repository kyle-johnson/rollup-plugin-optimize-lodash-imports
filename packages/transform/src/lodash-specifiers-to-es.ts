import type { ImportSpecifier } from "estree";

/**
 * Turns a generic lodash import into a specific import referencing the "lodash-es"
 * pacakge. Note that lodash-es cannot be imported from CommonJS.
 *
 * @param base "lodash" or "lodash/fp"
 * @param specifiers from an AST; assumes they are all ImportSpecifiers
 */
export function lodashSpecifiersToEs(
  base: string,
  specifiers: Array<ImportSpecifier>,
): Array<string> {
  const isFp = base.endsWith("fp");
  return specifiers.map(
    ({ imported, local }) =>
      `import { ${
        imported.name !== local.name
          ? imported.name + " as " + local.name
          : local.name
      } } from "lodash-es${isFp ? "/fp" : ""}";`,
  );
}
