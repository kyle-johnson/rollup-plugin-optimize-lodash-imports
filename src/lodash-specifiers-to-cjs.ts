import type { ImportSpecifier } from "estree";

/**
 * Turns a generic lodash import into a specific import using the CJS
 * lodash package.
 *
 * @param base "lodash" or "lodash/fp"
 * @param specifiers from an AST, assumes they are all ImportSpecifiers
 */
export function lodashSpecifiersToCjs(
  base: string,
  specifiers: Array<ImportSpecifier>
): Array<string> {
  return specifiers.map(
    ({ imported, local }) =>
      `import ${
        imported.name !== local.name ? local.name : imported.name
      } from "${base}/${imported.name}";`
  );
}
