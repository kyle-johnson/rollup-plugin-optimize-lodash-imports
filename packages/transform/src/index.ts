import type { Node } from "acorn";
import MagicString, { SourceMap } from "magic-string";
import { walk } from "estree-walker";

import {
  isImportDeclaration,
  isImportSpecifierArray,
  isProgram,
} from "./guards";
import { lodashSpecifiersToEs } from "./lodash-specifiers-to-es";
import { lodashSpecifiersToCjs } from "./lodash-specifiers-to-cjs";

// acorn adds these
declare module "estree" {
  interface BaseNodeWithoutComments {
    // added by acorn
    start: number;
    end: number;
  }
}

export type UNCHANGED = null;
export const UNCHANGED = null;

export interface CodeWithSourcemap {
  code: string;
  map: SourceMap;
}
export type ParseFunction = (code: string) => Node;
export type WarnFunction = (message: string) => void;

export function transform({
  code,
  id,
  parse,
  warn,
  useLodashEs,
  appendDotJs = true,
}: {
  code: string;
  id: string;
  parse: ParseFunction;
  warn?: WarnFunction;
  useLodashEs?: true;
  appendDotJs?: boolean;
}): CodeWithSourcemap | UNCHANGED {
  // before parsing, check if we can skip the whole file
  if (!code.includes("lodash")) {
    return UNCHANGED;
  }

  let ast;
  try {
    ast = parse(code);
  } catch (error) {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    (error as Error).message += ` in ${id}`;
    throw error;
  }

  // source map generation
  let magicString: MagicString | undefined;

  walk(ast, {
    enter(node) {
      // top-level node; we need to walk its children to find imports
      if (isProgram(node)) {
        return;
      }

      // skip any nodes that aren't imports (this skips most everything)
      if (!isImportDeclaration(node)) {
        this.skip();
        return;
      }

      // narrow-in on lodash imports we care about
      if (node.source.value !== "lodash" && node.source.value !== "lodash/fp") {
        this.skip();
        return;
      }

      // transform specific "lodash" and "lodash/fp" imports such as:
      // import { isNil } from "lodash";
      if (isImportSpecifierArray(node.specifiers)) {
        magicString = magicString ?? new MagicString(code);

        // modify
        const imports = useLodashEs
          ? lodashSpecifiersToEs(node.source.value, node.specifiers)
          : lodashSpecifiersToCjs(
              node.source.value,
              node.specifiers,
              appendDotJs
            );

        // write
        magicString.overwrite(node.start, node.end, imports.join("\n"));

        // no need to dig deeper
        this.skip();
      } else if (warn !== undefined) {
        // help end-users benefit from this plugin (this behavior differs from
        // babel-plugin-lodash which does optimize non-specific imports)
        warn(
          `Detected a default lodash or lodash/fp import within ${id} on line ${
            node.loc?.start?.line ?? "unknown"
          }.\nThis import cannot be optimized by optimize-lodash-imports.`
        );
      }
    },
  });

  if (!magicString) {
    return UNCHANGED;
  }

  return {
    code: magicString.toString(),
    map: magicString.generateMap({
      file: id,
      includeContent: true,
      hires: true,
    }),
  };
}
