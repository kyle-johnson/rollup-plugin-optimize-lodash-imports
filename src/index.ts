import type { Plugin } from "rollup";
import { createFilter, FilterPattern } from "@rollup/pluginutils";
import MagicString from "magic-string";
import { walk } from "estree-walker";

import {
  isImportDeclaration,
  isImportSpecifierArray,
  isProgram,
} from "./guards";
import { lodashSpecifiersToCjs } from "./lodash-specifiers-to-cjs";
import { lodashSpecifiersToEs } from "./lodash-specifiers-to-es";

declare module "estree" {
  interface BaseNodeWithoutComments {
    // added by acorn
    start: number;
    end: number;
  }
}

export type OptimizeLodashOptions = {
  /**
   * A minimatch pattern, or array of patterns, of files that should be
   * processed by this plugin (if omitted, all files are included by default)
   */
  include?: FilterPattern;
  /**
   * Files that should be excluded, if `include` is otherwise too permissive.
   */
  exclude?: FilterPattern;

  /**
   * Changes *all* lodash imports (but not lodash/fp imports!) to 'lodash-es' imports.
   * Don't use this for CommonJS outputs, the plugin will error should you do so.
   */
  useLodashEs?: boolean;
};

const UNCHANGED = null;

/**
 * Converts lodash imports to be specific, enabling better tree-shaking:
 *
 * `import { isNil } from "lodash";` -> `import { isNil } from "lodash/isNil";`
 *
 * Note that only specific named imports are supported, unlike babel-plugin-lodash. For example,
 * this plugin will print a warning for this import and make no changes to the import:
 *
 * `import _ from "lodash";`
 *
 * Optionally, set `useLodashEs` to true and `lodash` imports will be converted to `lodash-es`
 * imports. Note that it's up to user to include the `lodash-es` module and ensure the output
 * is set to some form of `es` (other output formats will error). An example:
 *
 * `import { isNil } from "lodash";` -> `import { isNil } from "lodash-es";`
 *
 * @param include files/globs to include with this plugin (optional)
 * @param exclude files/globs to exclude from this plugin (optional)
 * @param useLodashEs set `true` to convert imports to use "lodash-es" (optional; default false)
 */
export default function optimizeLodashImports({
  include,
  exclude,
  useLodashEs,
}: OptimizeLodashOptions = {}): Plugin & Required<Pick<Plugin, "transform">> {
  const filter = createFilter(include, exclude);

  return {
    name: "optimize-lodash-imports",
    outputOptions(options) {
      if (useLodashEs && options.format !== "es") {
        this.error(
          `'useLodashEs' is true but the output format is not 'es', it's ${
            options.format ?? "undefined"
          }`
        );
      }
      return UNCHANGED;
    },
    transform(code, id) {
      const warn = this.warn.bind(this);

      // sometimes we can skip the whole file
      if (!filter(id) || !code.includes("lodash")) {
        return UNCHANGED;
      }

      let ast;
      try {
        ast = this.parse(code);
      } catch (error) {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
        error.message += ` in ${id}`;
        throw error;
      }

      // easy source map generation
      let magicString: MagicString | undefined;

      walk(ast, {
        enter(node) {
          // top-level node; we need to walk its children to find imports
          if (isProgram(node)) {
            return;
          }

          // skip any nodes that aren't imports (this skips most everything)
          if (!isImportDeclaration(node)) {
            this.skip();
            return;
          }

          // narrow-in on lodash imports we care about
          if (
            node.source.value !== "lodash" &&
            node.source.value !== "lodash/fp"
          ) {
            this.skip();
            return;
          }

          // transform specific "lodash" and "lodash/fp" imports such as:
          // import { isNil } from "lodash";
          if (isImportSpecifierArray(node.specifiers)) {
            magicString = magicString ?? new MagicString(code);

            // modify
            const imports = useLodashEs
              ? lodashSpecifiersToEs(node.source.value, node.specifiers)
              : lodashSpecifiersToCjs(node.source.value, node.specifiers);

            // write
            magicString.overwrite(node.start, node.end, imports.join("\n"));

            // no need to dig deeper
            this.skip();
          } else {
            // help end-users benefit from this plugin (this behavior differs from
            // babel-plugin-lodash which does optimize non-specific imports)
            warn(
              `Detected a default lodash or lodash/fp import within ${id} on line ${
                node.loc?.start?.line ?? "unknown"
              }.\nThis import cannot be optimized by optimize-lodash-imports.`
            );
          }
        },
      });

      if (!magicString) {
        return UNCHANGED;
      }

      return {
        code: magicString.toString(),
        map: magicString.generateMap({
          file: id,
          includeContent: true,
          hires: true,
        }),
      };
    },
  };
}
