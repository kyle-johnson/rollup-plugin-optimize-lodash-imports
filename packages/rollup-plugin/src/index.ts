import type { Plugin } from "rollup";
import { createFilter, FilterPattern } from "@rollup/pluginutils";
import { transform as lodashTransform } from "@optimize-lodash/transform";

export type OptimizeLodashOptions = {
  /**
   * A minimatch pattern, or array of patterns, of files that should be
   * processed by this plugin (if omitted, all files are included by default)
   */
  include?: FilterPattern;
  /**
   * Files that should be excluded, if `include` is otherwise too permissive.
   */
  exclude?: FilterPattern;

  /**
   * Changes *all* lodash imports (but not lodash/fp imports!) to 'lodash-es' imports.
   * Don't use this for CommonJS outputs, the plugin will error should you do so.
   */
  useLodashEs?: true;
};

const UNCHANGED = null;

/**
 * Converts lodash imports to be specific, enabling better tree-shaking:
 *
 * `import { isNil } from "lodash";` -> `import { isNil } from "lodash/isNil";`
 *
 * Note that only specific named imports are supported, unlike babel-plugin-lodash. For example,
 * this plugin will print a warning for this import and make no changes to the import:
 *
 * `import _ from "lodash";`
 *
 * Optionally, set `useLodashEs` to true and `lodash` imports will be converted to `lodash-es`
 * imports. Note that it's up to user to include the `lodash-es` module and ensure the output
 * is set to some form of `es` (other output formats will error). An example:
 *
 * `import { isNil } from "lodash";` -> `import { isNil } from "lodash-es";`
 *
 * @param include files/globs to include with this plugin (optional)
 * @param exclude files/globs to exclude from this plugin (optional)
 * @param useLodashEs set `true` to convert imports to use "lodash-es" (optional; default false)
 */
export function optimizeLodashImports({
  include,
  exclude,
  useLodashEs,
}: OptimizeLodashOptions = {}): Plugin & Required<Pick<Plugin, "transform">> {
  const filter = createFilter(include, exclude);

  return {
    name: "optimize-lodash-imports",
    outputOptions(options) {
      if (useLodashEs && options.format !== "es") {
        this.error(
          `'useLodashEs' is true but the output format is not 'es', it's ${
            options.format ?? "undefined"
          }`
        );
      }
      return UNCHANGED;
    },
    transform(code, id) {
      const warn = this.warn.bind(this);
      const parse = this.parse.bind(this);

      // honor include/exclude
      if (!filter(id)) {
        return UNCHANGED;
      }

      return lodashTransform({ code, id, parse, warn, useLodashEs });
    },
  };
}
