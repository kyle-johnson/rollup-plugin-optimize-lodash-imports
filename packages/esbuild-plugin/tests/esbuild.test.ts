import path from "path";

import * as esbuild from "esbuild";

import { lodashOptimizeImports } from "../src";

describe("esbuild sanity check", () => {
  // esbuild is under ongoing development, so this test may give an indicator of
  // what might have changed when the change is with esbuild :-)
  test.each<[string]>([["no-transform.js"]])("%s", (filename) => {
    const result = esbuild.buildSync({
      entryPoints: [path.resolve(__dirname, "fixtures", filename)],
      sourcemap: false,
      write: false,
      format: "cjs",
      target: "es6",
    });

    expect(result.outputFiles).toHaveLength(1);
    expect(
      Buffer.from(result.outputFiles[0].contents).toString("utf-8")
    ).toMatchSnapshot();
  });
});

describe("esbuild with lodashOptimizeImports()", () => {
  test.each<[string]>([["standard-and-fp.js"]])("CJS: %s", async (filename) => {
    const result = await esbuild.build({
      entryPoints: [path.resolve(__dirname, "fixtures", filename)],
      sourcemap: false,
      write: false,
      format: "cjs",
      target: "es2020",
      plugins: [lodashOptimizeImports()],
    });

    expect(result.outputFiles).toHaveLength(1);
    const code = Buffer.from(result.outputFiles[0].contents).toString("utf-8");

    // ensure all imports became more specific
    expect(code).not.toMatch(/["']lodash["']/g);
    expect(code).not.toMatch(/["']lodash\/fp["']/g);

    expect(code).toMatchSnapshot();
  });

  test.each<[string]>([["standard-and-fp.js"]])("ESM: %s", async (filename) => {
    const result = await esbuild.build({
      entryPoints: [path.resolve(__dirname, "fixtures", filename)],
      sourcemap: false,
      write: false,
      format: "esm",
      target: "es2020",
      plugins: [lodashOptimizeImports({ useLodashEs: true })],
    });

    expect(result.outputFiles).toHaveLength(1);
    const code = Buffer.from(result.outputFiles[0].contents).toString("utf-8");

    // ensure all imports became more specific
    expect(code).not.toMatch(/["']lodash["']/g);
    expect(code).not.toMatch(/["']lodash\/fp["']/g);

    // ensure we have some lodash-es imports
    expect(code).toMatch(/["']lodash-es["']/g);

    expect(code).toMatchSnapshot();
  });
});
