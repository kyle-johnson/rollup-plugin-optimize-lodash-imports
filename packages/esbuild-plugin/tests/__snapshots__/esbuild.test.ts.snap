// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`esbuild sanity check no-transform.js 1`] = `
"\\"use strict\\";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === \\"object\\" || typeof from === \\"function\\") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, \\"__esModule\\", { value: true }), mod);
var no_transform_exports = {};
__export(no_transform_exports, {
  hello: () => hello
});
module.exports = __toCommonJS(no_transform_exports);
function hello(name) {
  return \`Hello \${name != null ? name : \\"World\\"}!\`;
}
"
`;

exports[`esbuild with lodashOptimizeImports() CJS: standard-and-fp.js 1`] = `
"\\"use strict\\";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === \\"object\\" || typeof from === \\"function\\") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, \\"default\\", { value: mod, enumerable: true }) : target, mod));
var __toCommonJS = (mod) => __copyProps(__defProp({}, \\"__esModule\\", { value: true }), mod);
var standard_and_fp_exports = {};
__export(standard_and_fp_exports, {
  isNonNilArray: () => isNonNilArray
});
module.exports = __toCommonJS(standard_and_fp_exports);
var import_isNil = __toESM(require(\\"lodash/isNil.js\\"));
var import_negate = __toESM(require(\\"lodash/negate.js\\"));
var import_every = __toESM(require(\\"lodash/fp/every.js\\"));
const everyNonNil = (0, import_every.default)((0, import_negate.default)(import_isNil.default));
function isNonNilArray(input) {
  return Array.isArray(input) && everyNonNil(input);
}
"
`;

exports[`esbuild with lodashOptimizeImports() ESM: standard-and-fp.js 1`] = `
"import { isNil } from \\"lodash-es\\";
import { negate } from \\"lodash-es\\";
import { every } from \\"lodash-es/fp\\";
const everyNonNil = every(negate(isNil));
function isNonNilArray(input) {
  return Array.isArray(input) && everyNonNil(input);
}
export {
  isNonNilArray
};
"
`;
